skimage/segmentation/morphsnakes.py

```diff
@@ -0,0 +1,451 @@
+# -*- coding: utf-8 -*-
+
+from itertools import cycle
+
+import numpy as np
+from scipy import ndimage as ndi
+
+__all__ = ['morph_acwe',
+           'morph_gac',
+           'gborders',
+           'circle_level_set',
+           'checkerboard_level_set',
+           'SIoIS',
+           'ISoSI',
+           'curvop'
+           ]
+
+
+class _fcycle(object):
+
+    def __init__(self, iterable):
+        """Call functions from the iterable each time it is called."""
+        self.funcs = cycle(iterable)
+
+    def __call__(self, *args, **kwargs):
+        f = next(self.funcs)
+        return f(*args, **kwargs)
+
+
+# SI and IS operators for 2D and 3D.
+_P2 = [np.eye(3),
+       np.array([[0, 1, 0]] * 3),
+       np.flipud(np.eye(3)),
+       np.rot90([[0, 1, 0]] * 3)]
+_P3 = [np.zeros((3, 3, 3)) for i in range(9)]
+
+_P3[0][:, :, 1] = 1
+_P3[1][:, 1, :] = 1
+_P3[2][1, :, :] = 1
+_P3[3][:, [0, 1, 2], [0, 1, 2]] = 1
+_P3[4][:, [0, 1, 2], [2, 1, 0]] = 1
+_P3[5][[0, 1, 2], :, [0, 1, 2]] = 1
+_P3[6][[0, 1, 2], :, [2, 1, 0]] = 1
+_P3[7][[0, 1, 2], [0, 1, 2], :] = 1
+_P3[8][[0, 1, 2], [2, 1, 0], :] = 1
+
+
+def SI(u):
+    """SI operator."""
+
+    if np.ndim(u) == 2:
+        P = _P2
+    elif np.ndim(u) == 3:
+        P = _P3
+    else:
+        raise ValueError("u has an invalid number of dimensions "
+                         "(should be 2 or 3)")
+
+    erosions = []
+    for P_i in P:
+        erosions.append(ndi.binary_erosion(u, P_i))
+
+    return np.array(erosions, dtype=np.int8).max(0)
+
+
+def IS(u):
+    """IS operator."""
+
+    if np.ndim(u) == 2:
+        P = _P2
+    elif np.ndim(u) == 3:
+        P = _P3
+    else:
+        raise ValueError("u has an invalid number of dimensions "
+                         "(should be 2 or 3)")
+
+    dilations = []
+    for P_i in P:
+        dilations.append(ndi.binary_dilation(u, P_i))
+
+    return np.array(dilations, dtype=np.int8).min(0)
+
+
+def SIoIS(u):
```

Comment:
Would it be possible to have a more meaningful name for this function? I guess it's an acronym, so maybe the full name instead?

Reply:
This one is tricky. 'o' is the composition operator. SIoIS is the composition of IS followed by SI, and "SIoIS" is the name we gave to that operator in the paper. I can call it sup_inf_inf_sup (and inf_sup_sup_inf for ISoSI), but I'm not sure if it will be clearer that way…

Reply:
Alternatively, since those functions are never called explicitly, I can remove them and rewrite `curvop` as

```python
curvop = _fcycle([lambda u: sup_inf(inf_sup(u)),   # SIoIS
                  lambda u: inf_sup(sup_inf(u))])  # ISoSI
```